core/src/main/java/org/apache/iceberg/BaseRewriteManifests.java

```diff
@@ -113,65 +119,98 @@ public RewriteManifests rewriteIf(Predicate<ManifestFile> pred) {
   }
 
   @Override
-  public List<ManifestFile> apply(TableMetadata base) {
-    Preconditions.checkNotNull(clusterByFunc, "clusterBy function cannot be null");
+  public RewriteManifests deleteManifest(ManifestFile manifest) {
+    deletedManifests.add(manifest);
+    return this;
+  }
+
+  @Override
+  public RewriteManifests addManifest(ManifestFile manifest) {
+    // the manifest must be rewritten with this update's snapshot ID
+    Map<Integer, PartitionSpec> specsById = ops.current().specsById();
+    try (ManifestReader reader = ManifestReader.read(ops.io().newInputFile(manifest.path()), specsById)) {
+      Set<ManifestEntry.Status> allowedStatuses = Sets.newHashSet(ManifestEntry.Status.EXISTING);
+      OutputFile newFile = manifestPath(manifestSuffix.getAndIncrement());
+      addedManifests.add(ManifestWriter.copyManifest(reader, newFile, snapshotId(), summaryBuilder, allowedStatuses));
+    } catch (IOException e) {
+      throw new RuntimeIOException(e, "Failed to close manifest: %s", manifest);
+    } catch (IllegalArgumentException e) {
+      throw new IllegalArgumentException("Cannot append the manifest: " + e.getMessage());
+    }
+    return this;
+  }
 
+  @Override
+  public List<ManifestFile> apply(TableMetadata base) {
     List<ManifestFile> currentManifests = base.currentSnapshot().manifests();
+    Set<ManifestFile> currentManifestSet = ImmutableSet.copyOf(currentManifests);
+
+    validateDeletedManifests(currentManifestSet);
 
-    if (requiresRewrite(currentManifests)) {
-      // run the rewrite process
+    if (requiresRewrite(currentManifestSet)) {
       performRewrite(currentManifests);
     } else {
-      // just keep any new manifests that were added since the last apply(), don't rerun
-      addExistingFromNewCommit(currentManifests);
```

Comment:
I think this comment and method name were useful to help understand what's happening here. We at least need to document that this function is responsible for adding manifests that were added since the snapshot that was originally rewritten.

Reply:
What about `keepNewActiveManifests`?

Reply:
The reason I modified this part is that the comment is no longer always true after this change. For example, when we use only add/delete manifests, there will be no rewrite and might be no retry, but we will still call `keepActiveManifests` to keep all active manifests.

Reply:
Got it, I think that works.